hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -14,30 +14,61 @@
 # END HEADER
 
 import math
-import re
-from typing import Any, Mapping, NamedTuple, Optional, Sequence, Tuple, Union
+from copy import deepcopy
+from typing import Any, Mapping, Optional, Sequence, Tuple, Union
 
 import numpy as np
 
-from hypothesis import assume, strategies as st
+from hypothesis import strategies as st
 from hypothesis.errors import InvalidArgument
+from hypothesis.extra._array_helpers import (
+    NDIM_MAX,
+    BasicIndex,
+    BasicIndexStrategy,
+    BroadcastableShapes,
+    Shape,
+    array_shapes,
+    broadcastable_shapes,
+    check_argument,
+    check_valid_dims,
+    mutually_broadcastable_shapes as _mutually_broadcastable_shapes,
+    order_check,
+    valid_tuple_axes as _valid_tuple_axes,
+)
 from hypothesis.internal.conjecture import utils as cu
 from hypothesis.internal.coverage import check_function
 from hypothesis.internal.reflection import proxies
-from hypothesis.internal.validation import check_type, check_valid_interval
+from hypothesis.internal.validation import check_type
 from hypothesis.strategies._internal.strategies import T, check_strategy
 from hypothesis.strategies._internal.utils import defines_strategy
-from hypothesis.utils.conventions import UniqueIdentifier, not_set
-
-Shape = Tuple[int, ...]
-# flake8 and mypy disagree about `ellipsis` (the type of `...`), and hence:
-BasicIndex = Tuple[Union[int, slice, "ellipsis", np.newaxis], ...]  # noqa: F821
-TIME_RESOLUTIONS = tuple("Y  M  D  h  m  s  ms  us  ns  ps  fs  as".split())
 
+__all__ = [
+    "Shape",
```

Comment:
Not documented as part of public api (sorry for leading you astray on this earlier!). We will keep `BroadcastableShapes`, but not the other types.

```suggestion
```